tuned_lens/scripts/ingredients.py

```diff
@@ -0,0 +1,215 @@
+        # by all ranks using MD5 hash of the command line arguments
+        if self.no_cache:
+            cache_dir = f"/tmp/{md5(pickle.dumps(self)).hexdigest()}"
+        else:
+            cache_dir = None
+
+        if self.random_model:
+            print(f"Sampling random weights for model '{self.name}'...")
+            config = AutoConfig.from_pretrained(self.name, revision=self.revision)
+            model = AutoModelForCausalLM.from_config(  # type: ignore
+                config, torch_dtype=th.float16
+            )
+        else:
+            print(f"Loading pretrained weights for '{self.name}'...")
+            model = AutoModelForCausalLM.from_pretrained(  # type: ignore
+                self.name,
+                cache_dir=cache_dir,
+                low_cpu_mem_usage=True,
+                revision=self.revision,
+                torch_dtype="auto",
+            )
+
+        # Make sure all ranks have the model
+        if cache_dir:
+            if dist.is_initialized():
+                dist.barrier()
+            if local_rank() == 0:
+                shutil.rmtree(cache_dir)
+
+        assert isinstance(model, PreTrainedModel)
+        model.eval()
+        model.requires_grad_(False)
+
+        return model, self.load_tokenizer()
+
+
+@dataclass
+class Sharding:
+    """Configuration for Fully Sharded Data Parallelism."""
+
+    fsdp: bool = field(action="store_true")
+    """Run the model with Fully Sharded Data Parallelism."""
+
+    cpu_offload: bool = field(action="store_true")
+    """Use CPU offloading. Must be combined with fsdp"""
+
+    def shard_model(
+        self, model: PreTrainedModel
+    ) -> FullyShardedDataParallel | PreTrainedModel:
+        """Shard the model using Fully Sharded Data Parallelism."""
+        th.cuda.set_device(local_rank())
+
+        if self.fsdp:
+            _, layers = get_transformer_layers(model)
+            layer_cls = type(layers[0])
+            print(f"Using '{layer_cls.__name__}' for transformer_auto_wrap_policy.")
+            return FullyShardedDataParallel(
+                model,
+                auto_wrap_policy=partial(
+                    transformer_auto_wrap_policy, transformer_layer_cls={layer_cls}
+                ),
+                cpu_offload=CPUOffload(offload_params=self.cpu_offload),
+                device_id=local_rank(),
+                # This turns out to be important for training speed
+                forward_prefetch=True,
+                mixed_precision=MixedPrecision(
+                    param_dtype=th.float16,
+                    reduce_dtype=th.float16,
+                    buffer_dtype=th.float16,
+                ),
+            )
+        elif self.cpu_offload:
+            raise ValueError("CPU offload requires FSDP.")
+        else:
+            model.to(local_rank())
+            return model
+
+    def shard_lens(self, lens: TunedLens) -> DDP | TunedLens:
+        """Shard a tuned lens using DistributedDataParallel."""
```

Comment:
This isn't actually "sharding" the lens; DDP just copies it to all ranks. We could try applying FSDP to the lens although I haven't done that for reasons I don't 100% remember. I think we could do it but it would require making a new wrapper class for both the model and the tuned lens (which returns like, the normal model output and the lens output at all layers) and then wrap that wrapper with `FullyShardedDataParallel`. But that's bad because it materializes all the logits for all the layers in VRAM at once (not to mention that under the hood FSDP would be copying those logits around a lot between ranks). So yeah I think I chose not to include the lens in FSDP for efficiency/memory reasons